### ISSUE TITLE
a efficient GRPO LOSS kernel by triton, reduce 46G memory

### ISSUE THREAD

**Issue by mdy666** [huggingface/open-r1#615]
When compute grpo loss, it requires a lot of memory, because it should save the `logits / temperature` and `log_softmax` for the backward. I write an efficient triton kernel. It can reduce 46G memory, when set bs=16 output_len=2048 and vocab_size=151936.

This is my code repo: https://github.com/mdy666/mdy_triton/tree/main/triton_kernels/grpo

```bash
# step 1
git clone https://github.com/mdy666/mdy_triton
cd mdy_triton/triton_kernels/grpo
pip install .

# step2
# The TRL version must be 0.16. Add the following to the first line of `grpo.py` in `open-r1`:
from triton_grpo_loss import trigger
```

The following picture illustrates the principle of the algorithm.
![Image](https://github.com/user-attachments/assets/e8884eec-d07d-4f07-a446-79c1261ba0c3)

**Comment by dignfei** [comment]
"Does the 'saving 46G memory' refer to 46G per GPU or 46G in total across 8 GPUs? Also, could you share the detailed hardware configuration and the model being used?"

**Comment by mdy666** [comment]
> "Does the 'saving 46G memory' refer to 46G per GPU or 46G in total across 8 GPUs? Also, could you share the detailed hardware configuration and the model being used?"

Each GPU reduce 46G compared with the origin trl code `bs=16, vocab_size=151936 and output_length=2048`. I push a new code, there is a file named `test_memory.py` compared with `LigerKernel`.
My GPU is H200. 
torch2.6-cuda12.6
triton3.2
